lib/pbench/server/cache_manager.py

```diff
@@ -173,6 +185,67 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
     )
 
 
+class Inventory:
+    """Encapsulate the tarfile TarFile object and file stream
+
+    This encapsulation allows cleaner downstream handling, so that we can close
+    both the extractfile file stream and the tarball object itself when we're
+    done. This eliminates interference with later operations.
+    """
+
+    def __init__(self, stream: IO[bytes], tarfile: Optional[tarfile.TarFile] = None):
+        """Construct an instance to track extracted inventory
+
+        This encapsulates many byte stream operations so that it can be used
+        as if it were a byte stream.
+
+        Args:
+            stream: the data stream of a specific tarball member
+            tarfile: the TarFile object
+        """
+        self.tarfile = tarfile
+        self.stream = stream
+
+    def close(self):
+        """Close both the byte stream and, if open, a tarfile object"""
+        self.stream.close()
+        if self.tarfile:
+            self.tarfile.close()
+
+    def getbuffer(self):
+        """Return the underlying byte buffer (used by send_file)"""
+        return self.getbuffer()
+
+    def read(self, size: int = -1, /) -> bytes:
+        """Encapsulate a read operation"""
+        return self.stream.read(size)
+
+    def readable(self) -> bool:
+        """Return the readable state of the stream"""
+        return self.stream.readable()
+
+    def seek(self, offset: int, whence: int = SEEK_SET, /) -> int:
+        """Allow setting the relative position in the stream"""
+        return self.stream.seek(offset, whence)
+
+    def __str__(self) -> str:
+        """Return a string representation"""
+        return f"<Stream {self.stream} from {self.tarfile}"
+
+    def __iter__(self):
+        """Allow iterating through lines in the buffer"""
+        self.stream.seek(0)
+        return self
```

Comment:
At the time I was kinda going crazy trying to figure out why Werkzeug was telling me the stream was closed when it started and I'd put this in hoping it might help. But it also seemed that this would enable two successive for loops to go through the stream twice, which could be convenient if not common, and I left it.

You're right, however, that it's an uncommon construct and I probably shouldn't.